spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.module.resolver;
+
+import java.io.File;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.maven.repository.internal.MavenRepositorySystemUtils;
+import org.eclipse.aether.DefaultRepositorySystemSession;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.artifact.DefaultArtifact;
+import org.eclipse.aether.connector.basic.BasicRepositoryConnectorFactory;
+import org.eclipse.aether.impl.DefaultServiceLocator;
+import org.eclipse.aether.repository.LocalRepository;
+import org.eclipse.aether.repository.RemoteRepository;
+import org.eclipse.aether.resolution.ArtifactRequest;
+import org.eclipse.aether.resolution.ArtifactResolutionException;
+import org.eclipse.aether.resolution.ArtifactResult;
+import org.eclipse.aether.spi.connector.RepositoryConnectorFactory;
+import org.eclipse.aether.spi.connector.transport.TransporterFactory;
+import org.eclipse.aether.transport.file.FileTransporterFactory;
+import org.eclipse.aether.transport.http.HttpTransporterFactory;
+
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.core.io.Resource;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An implementation of ModuleResolver using <a href="http://www.eclipse.org/aether/>aether</a> resolve the module
+ * artifact (uber jar) in a local Maven repository, downloading the latest update from a remote repository if
+ * necessary.
+ * @author David Turanski
+ */
+public class DefaultModuleResolver implements ModuleResolver {
+	private static final String DEFAULT_CONTENT_TYPE = "default";
+
+	private static final String DEFAULT_CLASSIFIER = "";
+
+	private static final String DEFAULT_EXTENSION = "jar";
+
+	private final File localRepository;
+
+	private final List<RemoteRepository> remoteRepositories;
+
+	private final RepositorySystem repositorySystem;
+
+	@Override
+	public Resource resolve(String groupId, String artifactId, String version) {
+		Artifact artifact = new DefaultArtifact(groupId, artifactId, DEFAULT_CLASSIFIER, DEFAULT_EXTENSION, version);
+		RepositorySystemSession session = newRepositorySystemSession(repositorySystem,
+				localRepository.getAbsolutePath());
+		ArtifactResult result;
+
+		try {
+			result = repositorySystem.resolveArtifact(session, new ArtifactRequest(artifact, remoteRepositories,
+					"runtime"));
+		}
+		catch (ArtifactResolutionException e) {
+			throw new RuntimeException(e);
```

Comment:
Perhaps we should introduce a `ModuleResolutionException`. We could pass the coordinates to the constructor and display those in the message (and still have the wrapped exception as the root cause), although in this case - which might be our only implementation - the wrapped Exception provides that info already. Another alternative would be `IllegalArgumentException` along with corresponding JavaDoc explaining that might happen if the coordinates are unresolvable. That said, I don't view this as an *illegal argument* per se, just one that is not resolvable.